### PR TITLE
Added Procedure Tests

### PR DESCRIPTION
**Pull request by BeksOmega**
## The basics

- [X] I branched from develop
- [X] My pull request is against develop
- [X] My code follows the [style guide](https://developers.google.com/blockly/guides/modify/web/style-guide)

### Resolves

I think this should cover closing #860 

### Proposed Changes

Adds tests to start covering procedure behavior.

[EDIT: I also removed the old jsunit tests because they were inaccurate]

### Reason for Changes

Procedures were dangerously undocumented.

### Test Coverage

Adds tests for:
* isNameUsed
* rename
* getCallers
* getDefinition
* Mutator / Arguments

### Additional Information

Some things are broken, those tests are skipped. Here are two particularly bad ones I took gifs of:
Procedures_WhitespaceThenText
Procedures_EmptyNew

But most things should be pretty easy to fix.

I'll create issues for the skipped tests once I get confirmation that all of them are bad.

### Additional Additional Information

I'd also like to create XML tests for the procedures, would you like that in this file, the xml file, a new file, or none file (if you don't think it's useful)?

### CODE REVIEW

File: blocks/procedures.js [hunk 1]
```diff
@@ -164,7 +164,7 @@ Blockly.Blocks['procedures_defnoreturn'] = {
    * @this Blockly.Block
    */
   decompose: function(workspace) {
-    var containerBlock = workspace.newBlock('procedures_mutatorcontainer');
+    /*var containerBlock = workspace.newBlock('procedures_mutatorcontainer');
```
Comment (BeksOmega) [inline blocks/procedures.js]: I had to change this because the initSvg() calls were breaking my tests.

File: blocks/procedures.js [hunk 2]
```diff
@@ -205,11 +234,7 @@ Blockly.Blocks['procedures_defnoreturn'] = {
       var varName = paramBlock.getFieldValue('NAME');
       this.arguments_.push(varName);
       var variable = this.workspace.getVariable(varName, '');
-      if (variable != null) {
```
Comment (BeksOmega) [inline blocks/procedures.js]: This was added by #2186 to fix #1958, but fields are happy now so this is unnecessary. Tests cover this change.

**Comment by BeksOmega** [comment]
Set the broken mocha test to skip because I'm not sure why the [dropdownCreate function](https://github.com/google/blockly/blob/2aa05c6d6d0373bd2c606ccfefd022710bc52be0/core/field_variable.js#L345) looks like this:

```
options.push([Blockly.Msg['RENAME_VARIABLE'], Blockly.RENAME_VARIABLE_ID]);
  if (Blockly.Msg['DELETE_VARIABLE']) {
    options.push(
        [
          Blockly.Msg['DELETE_VARIABLE'].replace('%1', name),
          Blockly.DELETE_VARIABLE_ID
        ]
    );
  }
```

It broke because I had to add the msg files to the mocha and it started adding the delete option.

**Comment by alschmiedt** [comment]
XML tests would be great. If you are planning on writing more than 2 or 3 tests then I would create a xml_procedures_test.js file, otherwise I would just put them into the xml_test.js file

File: tests/mocha/field_variable_test.js [hunk 3]
```diff
@@ -68,7 +68,7 @@ suite('Variable Fields', function() {
     sinon.restore();
   });
 
-  test('Dropdown contains variables', function() {
+  test.skip('Dropdown contains variables', function() {
```
Comment (alschmiedt) [inline tests/mocha/field_variable_test.js]: I think the move here is to update the test to expect 5 instead of 4 since having the delete option is expected behavior.

Comment (BeksOmega) [reply tests/mocha/field_variable_test.js]: Fixed.

File: blocks/procedures.js [hunk 4]
```diff
@@ -185,7 +185,36 @@ Blockly.Blocks['procedures_defnoreturn'] = {
       paramBlock.oldLocation = i;
       connection.connect(paramBlock.previousConnection);
       connection = paramBlock.nextConnection;
+    }*/
+
```
[the remark this answers is not in the document]
Comment (BeksOmega) [reply blocks/procedures.js]: Fixed.

File: tests/mocha/procedures_test.js [hunk 5]
```diff
@@ -0,0 +1,421 @@
...
+      test.skip('CAPS -> lower', function() {
+        this.callForAllTypes(function() {
+          createMutator.call(this, ['ARG']);
+          this.argBlock.setFieldValue('arg', 'NAME');
+          this.defBlock.compose(this.containerBlock);
+          assertArgs.call(this, ['arg']);
+          clearVariables.call(this);
+        }, 'name');
+      });
+      test('Set Arg Empty (#1958)', function() {
```
[the remark this answers is not in the document]
Comment (BeksOmega) [reply tests/mocha/procedures_test.js]: Fixed.

File: tests/mocha/procedures_test.js [hunk 6]
```diff
@@ -0,0 +1,421 @@
...
+      }, 'name1');
+    });
+    // This can happen using the trashcan to retrieve blocks.
+    test.skip('Call Different Case', function() {
+      this.callForAllTypes(function() {
+        this.callBlock.setFieldValue('NAME', 'NAME');
+        var callers = Blockly.Procedures.getCallers('name1', this.workspace);
+        chai.assert.equal(callers.length, 1);
+        chai.assert.equal(callers[0], this.callBlock);
+      }, 'name');
```
[the remark this answers is not in the document]
Comment (BeksOmega) [reply tests/mocha/procedures_test.js]: I wrote down a procedure for how to replicate it, does it sound good to you?

File: tests/mocha/procedures_test.js [hunk 7]
```diff
@@ -0,0 +1,421 @@
...
+
+      for (var i = 0, types; types = typesArray[i]; i++) {
+        var context = Object.create(null);
+        context.workspace = this.workspace;
+        context.defType = types[0];
+        context.callType = types[1];
+
+        context.defBlock = new Blockly.Block(this.workspace, types[0]);
+        context.defBlock.setFieldValue(startName, 'NAME');
+        context.callBlock = new Blockly.Block(this.workspace, types[1]);
```
[the remark this answers is not in the document]
Comment (BeksOmega) [reply tests/mocha/procedures_test.js]: Fixed

File: tests/mocha/procedures_test.js [hunk 8]
```diff
@@ -0,0 +1,421 @@
...
+        ['procedures_defreturn', 'procedures_callreturn']
+      ];
+
+      for (var i = 0, types; types = typesArray[i]; i++) {
+        var context = Object.create(null);
+        context.workspace = this.workspace;
+        context.defType = types[0];
+        context.callType = types[1];
+
+        context.defBlock = new Blockly.Block(this.workspace, types[0]);
```
[the remark this answers is not in the document]
Comment (BeksOmega) [reply tests/mocha/procedures_test.js]: Fixed

File: tests/mocha/index.html [hunk 9]
```diff
@@ -6,6 +6,16 @@
 
     <link href="https://unpkg.com/mocha@5.2.0/mocha.css" rel="stylesheet" />
     <script src="../../blockly_uncompressed.js"></script>
+    <script src="../../msg/messages.js"></script>
```
[the remark this answers is not in the document]
Comment (BeksOmega) [reply tests/mocha/index.html]: Removed

File: tests/mocha/procedures_test.js [hunk 10]
```diff
@@ -0,0 +1,421 @@
+/**
+ * @license
+ * Blockly Tests
+ *
+ * Copyright 2017 Google Inc.
```
[the remark this answers is not in the document]
Comment (BeksOmega) [reply tests/mocha/procedures_test.js]: Fixed

**Comment by BeksOmega** [comment]
> XML tests would be great. If you are planning on writing more than 2 or 3 tests then I would create a xml_procedures_test.js file, otherwise I would just put them into the xml_test.js file

Noted, probably won't get to it today but I have put it in my task management system.

File: tests/mocha/procedures_test.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (alschmiedt) [reply tests/mocha/procedures_test.js]: Yup that is great thanks!

**Comment by BeksOmega** [comment]
@alschmiedt Did you want me to create issues for those skipped tests?

**Comment by alschmiedt** [comment]
Yup that would be great : ) I looked over them and they all seem like valid issues.
